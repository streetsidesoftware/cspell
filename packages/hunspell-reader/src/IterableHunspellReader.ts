import * as fs from 'fs/promises';
import type { Sequence } from 'gensequence';
import { genSequence } from 'gensequence';
import { decode } from 'iconv-lite';

import type { Aff } from './aff';
import type { AffWord } from './affDef';
import { parseAffFileToAff } from './affReader';
import type { WordInfo } from './types';
import { filterOrderedList } from './util';

const defaultEncoding = 'UTF-8';

export { WordInfo } from './types';

export interface HunspellSrcData {
    /** The Aff rules to use with the dictionary entries */
    aff: Aff;
    /** the hunspell dictionary entries complete with affix flags */
    dic: string[];
}

export class IterableHunspellReader implements Iterable<string> {
    readonly aff: Aff;

    constructor(readonly src: HunspellSrcData) {
        this.aff = src.aff;
    }

    get dic() {
        return this.src.dic;
    }

    set maxDepth(value: number) {
        this.aff.maxSuffixDepth = value;
    }

    get maxDepth(): number {
        return this.aff.maxSuffixDepth;
    }

    /** the number of .dic entries */
    get size(): number {
        return this.src.dic.length;
    }

    /**
     * @internal
     */
    dicWordsSeq(): Sequence<WordInfo> {
        return genSequence(this.src.dic).map((line) => {
            const [word, rules] = line.split('/', 2);
            return { word, rules, prefixes: [], suffixes: [] };
        });
    }

    /**
     * iterates through the root words of the dictionary
     */
    iterateRootWords(): Iterable<string> {
        return this.seqRootWords();
    }

    iterateWords(): Iterable<string> {
        return this.seqWords();
    }

    /**
     * Iterator for all the words in the dictionary. The words are in the order found in the .dic after the
     * transformations have been applied. Forbidden and CompoundOnly words are filtered out.
     */
    [Symbol.iterator]() {
        return this.wholeWords();
    }

    /**
     * create an iterable sequence of the words in the dictionary.
     *
     * @param tapPreApplyRules -- optional function to be called before rules are applied to a word.
     *                            It is mostly used for monitoring progress in combination with `size`.
     */
    seqAffWords(tapPreApplyRules?: (dicEntry: string, index: number) => void, maxDepth?: number) {
        return this.seqTransformDictionaryEntries(tapPreApplyRules, maxDepth).concatMap((a) => a);
    }

    /**
     * create an iterable sequence of the words in the dictionary.
     *
     * @param tapPreApplyRules -- optional function to be called before rules are applied to a word.
     *                            It is mostly used for monitoring progress in combination with `size`.
     */
    seqTransformDictionaryEntries(
        tapPreApplyRules?: (dicEntry: string, index: number) => void,
        maxDepth?: number
    ): Sequence<AffWord[]> {
        const seq = genSequence(this.src.dic);
        let count = 0;
        const dicWords = tapPreApplyRules ? seq.map((a) => (tapPreApplyRules(a, count++), a)) : seq;
        return dicWords.map((dicWord) => this.aff.applyRulesToDicEntry(dicWord, maxDepth));
    }

    /**
     * Iterator for all the words in the dictionary. The words are in the order found in the .dic after the
     * transformations have been applied. Forbidden and CompoundOnly ARE INCLUDED.
     *
     * @internal
     */
    seqWords() {
        return this.seqAffWords()
            .map((w) => w.word)
            .filter(createMatchingWordsFilter());
    }

    /**
     * Returns an iterable that will only return stand alone words.
     */
    wholeWords() {
        return (
            this.seqAffWords()
                // Filter out words that are forbidden or only allowed in Compounds.
                .filter((w) => !w.flags.isForbiddenWord && !w.flags.isOnlyAllowedInCompound)
                .map((w) => w.word)
                .filter(createMatchingWordsFilter())
        );
    }

    /**
     * @internal
     */
    seqRootWords() {
        return this.dicWordsSeq().map((w) => w.word);
    }

    /**
     *
     * @param affFile - path to aff file.
     * @param dicFile - path to dic file.
     * @returns IterableHunspellReader
     */
    static async createFromFiles(affFile: string, dicFile: string) {
        const aff = await parseAffFileToAff(affFile, defaultEncoding);
        const buffer = await fs.readFile(dicFile);
        const dicFileContent = decode(buffer, aff.affInfo.SET);
        const dic = dicFileContent
            .split('\n')
            .slice(1) // The first entry is the count of entries.
            .map((a) => a.trim())
            .filter((line) => !!line);
        return new IterableHunspellReader({ aff, dic });
    }
}

export function createMatchingWordsFilter() {
    return filterOrderedList<string>((a, b) => a !== b);
}
