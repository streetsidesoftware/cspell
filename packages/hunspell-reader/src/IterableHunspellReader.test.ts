import { readdirSync } from 'fs';
import { genSequence } from 'gensequence';
import * as path from 'path';
import { describe, expect, it, test } from 'vitest';

import * as Aff from './aff.js';
import * as AffReader from './affReader.js';
import { IterableHunspellReader } from './IterableHunspellReader.js';

const DICTIONARY_LOCATIONS = path.join(__dirname, '..', 'dictionaries');

describe('Basic Validation of the Reader', () => {
    const pSimpleAff = getSimpleAff();

    it('Validate the dictionary', async () => {
        const aff = await pSimpleAff;
        const src = { aff, dic: textToArray(simpleWords) };
        const reader = new IterableHunspellReader(src);
        expect(reader.dic).toEqual(textToArray(simpleWords));
    });

    it('Validate Simple Words List', async () => {
        const aff = await pSimpleAff;
        const src = { aff, dic: textToArray(simpleWords) };
        const reader = new IterableHunspellReader(src);
        expect([...reader.iterateRootWords()]).toEqual(['happy', 'ring']);
        const tapped: string[] = [];
        const words = [...reader.seqAffWords((w) => tapped.push(w)).map((a) => a.word)];
        const someExpectedWords = ['happy', 'unhappy', 'happily', 'unhappily'];
        expect(words).toEqual(expect.arrayContaining(someExpectedWords));
        expect(tapped).toEqual(['happy/UY', 'ring/AUJ']);
    });

    it('Validate Expanding `place/AJ`', async () => {
        const aff = await pSimpleAff;
        const src = { aff, dic: ['place/AJ'] };
        const reader = new IterableHunspellReader(src);
        // cspell:ignore replacings
        expect([...reader]).toEqual(['place', 'placing', 'placings', 'replace', 'replacing', 'replacings']);
        expect(reader.size).toBe(1);
    });

    it('Validates prefix/suffix/base', async () => {
        const aff = await pSimpleAff;
        const src = { aff, dic: textToArray(simpleWords) };
        const reader = new IterableHunspellReader(src);
        const results = [...reader.seqAffWords().map((a) => a.prefix + '<' + a.base + '>' + a.suffix)];
        const someExpectedWords = ['<happy>', 'un<happy>', '<happ>ily', 'un<happ>ily']; // cspell:ignore happ
        expect(results).toEqual(expect.arrayContaining(someExpectedWords));
    });

    it('sets the max depth', async () => {
        const aff = await pSimpleAff;
        const src = { aff, dic: textToArray(simpleWords) };
        const reader = new IterableHunspellReader(src);
        const depth = reader.maxDepth;
        reader.maxDepth = 0;
        const results = [...reader.seqAffWords().map((a) => a.prefix + '<' + a.base + '>' + a.suffix)].sort();
        const expectedWords = ['<happy>', '<ring>'].sort(); // cspell:ignore happ
        reader.maxDepth = depth;
        expect(results).toEqual(expectedWords);
    });

    it('Iterates a few words', async () => {
        const aff = await pSimpleAff;
        const src = { aff, dic: textToArray(simpleWords) };
        const reader = new IterableHunspellReader(src);
        const words = [...reader.iterateWords()];
        expect(words).toEqual(expect.arrayContaining(['happy', 'unhappy', 'happily', 'unhappily', 'ring']));
    });
});

const timeout = 10000;

describe('HunspellReader En', function () {
    // We are reading big files, so we need to give it some time.
    const aff = __dirname + '/../dictionaries/en_US.aff';
    const dic = __dirname + '/../dictionaries/en_US.dic';
    const pReader = IterableHunspellReader.createFromFiles(aff, dic);

    it(
        'reads dict entries',
        async () => {
            const reader = await pReader;
            const values = reader.dicWordsSeq().skip(10000).take(10).toArray();
            expect(values.length).toBe(10);
        },
        timeout
    );

    it(
        'reads words with info',
        async () => {
            const reader = await pReader;
            const values = reader.seqWords().skip(10000).take(10).toArray();
            expect(values.length).toBe(10);
        },
        timeout
    );

    it(
        'reads words',
        async () => {
            const reader = await pReader;
            const values = genSequence(reader).skip(10000).take(10).toArray();
            expect(values.length).toBe(10);
        },
        timeout
    );

    it(
        'tests iterating through dictionary entry transformations',
        async () => {
            const reader = await pReader;
            const dicEntries: { word: string; index: number }[] = [];

            function recordEntries(word: string, index: number) {
                dicEntries.push({ word, index });
            }

            const values = reader.seqTransformDictionaryEntries(recordEntries, 2).skip(100).take(10).toArray();
            // Note the current implementation of Sequence will pre-fetch the next iteration
            // causing 1 more than expected to be recorded.
            expect(dicEntries).toHaveLength(111);
            expect(values[1][0].dic).toBe(dicEntries[101].word);
        },
        timeout
    );
});

describe('HunspellReader read dictionaries', function () {
    // We are reading big files, so we need to give it some time.

    test.each`
        hunDic
        ${'da_DK'}
        ${'nl'}
        ${'Portuguese (Brazilian)'}
        ${'en_US'}
    `(
        'reads words with info from "$hunDic"',
        async ({ hunDic }) => {
            const aff = __dirname + `/../dictionaries/${hunDic}.aff`;
            const dic = __dirname + `/../dictionaries/${hunDic}.dic`;
            const reader = await IterableHunspellReader.createFromFiles(aff, dic);
            const values = reader.seqWords().skip(200).take(200).toArray();
            expect(values.length).toBe(200);
            expect(values).toMatchSnapshot();
        },
        timeout
    );
});

describe('Validated loading all dictionaries in the `dictionaries` directory.', () => {
    const dictionaries = readdirSync(DICTIONARY_LOCATIONS)
        .filter((dic) => !!dic.match(/\.aff$/))
        .map((base) => path.join(DICTIONARY_LOCATIONS, base));
    it('Make sure we found some sample dictionaries', () => {
        expect(dictionaries.length).toBeGreaterThan(4);
    });

    dictionaries.forEach((dicAff) => {
        const dicDic = dicAff.replace(/\.aff$/, '.dic');
        it(`Ensure we can load aff ${path.basename(dicAff)}`, async () => {
            const aff = await AffReader.parseAffFile(dicAff);
            expect(aff.PFX).toBeInstanceOf(Map);
            expect(aff.SFX).toBeInstanceOf(Map);
        });

        it(`Ensure we can load the dictionary ${path.basename(dicDic)}`, async () => {
            const reader = await IterableHunspellReader.createFromFiles(dicAff, dicDic);
            const sample = reader.seqWords().take(100).toArray();
            expect(sample).toHaveLength(100);
        });
    });
});

function textToArray(text: string) {
    return text
        .split('\n')
        .filter((a) => !!a)
        .slice(1);
}

function getSimpleAff() {
    // cspell:ignore esianrtolcdugmphbyfvkwzESIANRTOLCDUGMPHBYFVKWZ
    const sampleAff = `
SET UTF-8
TRY esianrtolcdugmphbyfvkwzESIANRTOLCDUGMPHBYFVKWZ'
ICONV 1
ICONV â€™ '
NOSUGGEST !

# ordinal numbers
COMPOUNDMIN 1
# only in compounds: 1th, 2th, 3th
ONLYINCOMPOUND c
# compound rules:
# 1. [0-9]*1[0-9]th (10th, 11th, 12th, 56714th, etc.)
# 2. [0-9]*[02-9](1st|2nd|3rd|[4-9]th) (21st, 22nd, 123rd, 1234th, etc.)
COMPOUNDRULE 2
COMPOUNDRULE n*1t
COMPOUNDRULE n*mp
WORDCHARS 0123456789

PFX A Y 1
PFX A   0     re         .

PFX I Y 1
PFX I   0     in         .

PFX U Y 1
PFX U   0     un         .

PFX X Y 1
PFX X   0     un         .
PFX X   0     re         .
PFX X   0     in         .
PFX X   0     a          .

SFX Y Y 2
SFX Y   0     ly         [^y]
SFX Y   y     ily        [y]

SFX G Y 2
SFX G   e     ing        e
SFX G   0     ing        [^e]

SFX J Y 2
SFX J   e     ing/S      e
SFX J   0     ing/S      [^e]

SFX S Y 4
SFX S   y     ies        [^aeiou]y
SFX S   0     s          [aeiou]y
SFX S   0     es         [sxzh]
SFX S   0     s          [^sxzhy]

`;

    return new Aff.Aff(AffReader.parseAff(sampleAff));
}

const simpleWords = `
2
happy/UY
ring/AUJ
`;

// cspell:ignore moderne avoir huis pannenkoek ababillar CDSG ings AUGJ aeiou sxzh sxzhy
// cspell:words COMPOUNDMIN ONLYINCOMPOUND COMPOUNDRULE WORDCHARS
