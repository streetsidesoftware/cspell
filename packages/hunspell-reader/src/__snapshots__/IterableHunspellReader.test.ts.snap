// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HunspellReader read dictionaries reads words with info from "Portuguese (Brazilian)" 1`] = `
Array [
  "ab-reagindo-me",
  "ab-reagindo-nos",
  "ab-reagindo-o",
  "ab-reagindo-os",
  "ab-reagindo-se",
  "ab-reagindo-se-lhe",
  "ab-reagindo-se-lhes",
  "ab-reagindo-te",
  "ab-reagindo-vos",
  "ab-reagir",
  "ab-reagir-lhe",
  "ab-reagir-lhe-ei",
  "ab-reagir-lhe-emos",
  "ab-reagir-lhe-ia",
  "ab-reagir-lhe-iam",
  "ab-reagir-lhe-á",
  "ab-reagir-lhe-ão",
  "ab-reagir-lhes",
  "ab-reagir-lhes-ei",
  "ab-reagir-lhes-emos",
  "ab-reagir-lhes-ia",
  "ab-reagir-lhes-iam",
  "ab-reagir-lhes-á",
  "ab-reagir-lhes-ão",
  "ab-reagir-me",
  "ab-reagir-me-ei",
  "ab-reagir-me-ia",
  "ab-reagir-me-iam",
  "ab-reagir-me-á",
  "ab-reagir-me-ão",
  "ab-reagir-nos",
  "ab-reagir-nos-ei",
  "ab-reagir-nos-emos",
  "ab-reagir-nos-ia",
  "ab-reagir-nos-iam",
  "ab-reagir-nos-á",
  "ab-reagir-nos-ão",
  "ab-reagir-se",
  "ab-reagir-se-ei",
  "ab-reagir-se-ia",
  "ab-reagir-se-iam",
  "ab-reagir-se-lhe",
  "ab-reagir-se-lhe-á",
  "ab-reagir-se-lhes",
  "ab-reagir-se-lhes-ão",
  "ab-reagir-se-á",
  "ab-reagir-se-ão",
  "ab-reagir-te",
  "ab-reagir-te-ei",
  "ab-reagir-te-emos",
  "ab-reagir-te-ia",
  "ab-reagir-te-iam",
  "ab-reagir-te-á",
  "ab-reagir-te-ão",
  "ab-reagir-vos",
  "ab-reagir-vos-ei",
  "ab-reagir-vos-emos",
  "ab-reagir-vos-ia",
  "ab-reagir-vos-iam",
  "ab-reagir-vos-á",
  "ab-reagir-vos-ão",
  "ab-reagira",
  "ab-reagira-a",
  "ab-reagira-as",
  "ab-reagira-lhe",
  "ab-reagira-lhes",
  "ab-reagira-me",
  "ab-reagira-nos",
  "ab-reagira-o",
  "ab-reagira-os",
  "ab-reagira-se",
  "ab-reagira-te",
  "ab-reagira-vos",
  "ab-reagiram",
  "ab-reagiram-lhe",
  "ab-reagiram-lhes",
  "ab-reagiram-me",
  "ab-reagiram-na",
  "ab-reagiram-nas",
  "ab-reagiram-no",
  "ab-reagiram-nos",
  "ab-reagiram-se",
  "ab-reagiram-te",
  "ab-reagiram-vos",
  "ab-reagiras",
  "ab-reagirdes",
  "ab-reagirei",
  "ab-reagireis",
  "ab-reagirem",
  "ab-reagirem-lhe",
  "ab-reagirem-lhes",
  "ab-reagirem-me",
  "ab-reagirem-na",
  "ab-reagirem-nas",
  "ab-reagirem-no",
  "ab-reagirem-nos",
  "ab-reagirem-se",
  "ab-reagirem-te",
  "ab-reagirem-vos",
  "ab-reagiremos",
  "ab-reagires",
  "ab-reagiria",
  "ab-reagiriam",
  "ab-reagirias",
  "ab-reagirmo-nos",
  "ab-reagirmo-vos",
  "ab-reagirmos",
  "ab-reagirmos-lhe",
  "ab-reagirmos-lhes",
  "ab-reagirmos-te",
  "ab-reagirá",
  "ab-reagirás",
  "ab-reagirão",
  "ab-reagiríamo-nos",
  "ab-reagiríamo-vos",
  "ab-reagiríamos",
  "ab-reagiríamos-lhe",
  "ab-reagiríamos-lhes",
  "ab-reagiríamos-te",
  "ab-reagiríeis",
  "ab-reagis",
  "ab-reagisse",
  "ab-reagisse-a",
  "ab-reagisse-as",
  "ab-reagisse-lhe",
  "ab-reagisse-lhes",
  "ab-reagisse-me",
  "ab-reagisse-nos",
  "ab-reagisse-o",
  "ab-reagisse-os",
  "ab-reagisse-se",
  "ab-reagisse-te",
  "ab-reagisse-vos",
  "ab-reagissem",
  "ab-reagissem-lhe",
  "ab-reagissem-lhes",
  "ab-reagissem-me",
  "ab-reagissem-na",
  "ab-reagissem-nas",
  "ab-reagissem-no",
  "ab-reagissem-nos",
  "ab-reagissem-te",
  "ab-reagissem-vos",
  "ab-reagisses",
  "ab-reagiste",
  "ab-reagiste-a",
  "ab-reagiste-as",
  "ab-reagiste-lhe",
  "ab-reagiste-lhes",
  "ab-reagiste-me",
  "ab-reagiste-nos",
  "ab-reagiste-o",
  "ab-reagiste-os",
  "ab-reagiste-se",
  "ab-reagiste-te",
  "ab-reagiste-vos",
  "ab-reagistes",
  "ab-reagiu",
  "ab-reagiu-a",
  "ab-reagiu-as",
  "ab-reagiu-lhe",
  "ab-reagiu-lhes",
  "ab-reagiu-me",
  "ab-reagiu-nos",
  "ab-reagiu-o",
  "ab-reagiu-os",
  "ab-reagiu-se",
  "ab-reagiu-te",
  "ab-reagiu-vos",
  "ab-reagíamo-nos",
  "ab-reagíamo-vos",
  "ab-reagíamos",
  "ab-reagíamos-lhe",
  "ab-reagíamos-lhes",
  "ab-reagíamos-te",
  "ab-reagíeis",
  "ab-reagíramos",
  "ab-reagíreis",
  "ab-reagísseis",
  "ab-reagíssemos",
  "ab-reaja",
  "ab-reaja-a",
  "ab-reaja-as",
  "ab-reaja-lhe",
  "ab-reaja-lhes",
  "ab-reaja-me",
  "ab-reaja-nos",
  "ab-reaja-o",
  "ab-reaja-os",
  "ab-reaja-se",
  "ab-reaja-te",
  "ab-reaja-vos",
  "ab-reajais",
  "ab-reajam",
  "ab-reajam-lhe",
  "ab-reajam-lhes",
  "ab-reajam-me",
  "ab-reajam-na",
  "ab-reajam-nas",
  "ab-reajam-no",
]
`;

exports[`HunspellReader read dictionaries reads words with info from "da_DK" 1`] = `
Array [
  "Abdelkader",
  "Abdelkaders",
  "Abdelkarim",
  "Abdelkarims",
  "Abdellah",
  "Abdellahs",
  "Abdellatif",
  "Abdellatifs",
  "Abdelqadeem",
  "Abdelqadeems",
  "Abdelshakur",
  "Abdelshakurs",
  "Abderrahmane",
  "Abderrahmanes",
  "Abdeslam",
  "Abdeslams",
  "Abdi",
  "Abdis",
  "Abdiadan",
  "Abdiadans",
  "Abdiaziz",
  "Abdiaziz'",
  "Abdibaari",
  "Abdibaaris",
  "abdicer",
  "abdicere",
  "abdicerede",
  "abdiceredes",
  "abdicerende",
  "abdicerendes",
  "abdicerer",
  "abdiceres",
  "abdiceret",
  "abdicerets",
  "-abdicering",
  "-abdiceringen",
  "-abdiceringens",
  "-abdiceringer",
  "-abdiceringerne",
  "-abdiceringernes",
  "-abdiceringers",
  "-abdicerings",
  "abdicering",
  "abdiceringen",
  "abdiceringens",
  "abdiceringer",
  "abdiceringerne",
  "abdiceringernes",
  "abdiceringers",
  "abdicerings",
  "Abdifatah",
  "Abdifatahs",
  "Abdighani",
  "Abdighanis",
  "Abdihakin",
  "Abdihakins",
  "Abdikarim",
  "Abdikarims",
  "abdikation",
  "abdikationen",
  "abdikationens",
  "abdikationer",
  "abdikationerne",
  "abdikationernes",
  "abdikationers",
  "abdikations",
  "Abdikhaliq",
  "Abdikhaliqs",
  "Abdikheyr",
  "Abdikheyrs",
  "Abdillahi",
  "Abdillahis",
  "Abdimalik",
  "Abdimaliks",
  "Abdinasir",
  "Abdinasirs",
  "Abdiqafar",
  "Abdiqafars",
  "Abdiqani",
  "Abdiqanis",
  "Abdirahman",
  "Abdirahmans",
  "Abdirashid",
  "Abdirashids",
  "Abdirazaaq",
  "Abdirazaaqs",
  "Abdirisak",
  "Abdirisaks",
  "Abdirizak",
  "Abdirizaks",
  "Abdisalan",
  "Abdisalans",
  "Abdiwali",
  "Abdiwalis",
  "Abdol",
  "Abdols",
  "Abdollah",
  "Abdollahs",
  "-abdomen",
  "-abdomens",
  "abdomen",
  "abdomens",
  "Abdoulkader",
  "Abdoulkaders",
  "abducer",
  "abducere",
  "abducerede",
  "abduceredes",
  "abducerende",
  "abducerendes",
  "abducerer",
  "abduceres",
  "abduceret",
  "abducerets",
  "abduktion",
  "abduktionen",
  "abduktionens",
  "abduktions",
  "Abdul",
  "Abduls",
  "Abdulah",
  "Abdulahs",
  "Abdulahi",
  "Abdulahis",
  "Abdulaziz",
  "Abdulaziz'",
  "Abdulbasid",
  "Abdulbasids",
  "Abdulgadir",
  "Abdulgadirs",
  "Abdulkadir",
  "Abdulkadirs",
  "Abdulkarim",
  "Abdulkarims",
  "Abdulkhaliq",
  "Abdulkhaliqs",
  "Abdulla",
  "Abdullas",
  "Abdullah",
  "Abdullahs",
  "Abdullahi",
  "Abdullahis",
  "Abdulle",
  "Abdulles",
  "Abdullaahi",
  "Abdullaahis",
  "Abdulqadir",
  "Abdulqadirs",
  "Abdulrahman",
  "Abdulrahmans",
  "Abdurahman",
  "Abdurahmans",
  "Abdurrahman",
  "Abdurrahmans",
  "-ab",
  "-abe",
  "-abede",
  "-abedes",
  "-abende",
  "-abendes",
  "-aber",
  "-abes",
  "-abet",
  "-abets",
  "ab",
  "abe",
  "abede",
  "abedes",
  "abende",
  "abendes",
  "aber",
  "abes",
  "abet",
  "abets",
  "-abe",
  "-aben",
  "-abens",
  "-abenser",
  "-abenserne",
  "-abensernes",
  "-abensers",
  "-abenset",
  "-abensets",
  "-aber",
  "-aberne",
  "-abernes",
  "-abernes'",
  "-abers",
  "-abes",
  "abe",
  "aben",
  "abens",
  "abenser",
  "abenserne",
  "abensernes",
  "abensers",
  "abenset",
  "abensets",
  "aber",
  "aberne",
]
`;

exports[`HunspellReader read dictionaries reads words with info from "en_US" 1`] = `
Array [
  "Ada's",
  "Adas",
  "Adam",
  "Adam's",
  "Adams",
  "Adams's",
  "Adan",
  "Adan's",
  "Adana",
  "Adana's",
  "Adar",
  "Adar's",
  "Addams",
  "Addams's",
  "Adderley",
  "Adderley's",
  "Addie",
  "Addie's",
  "Addison",
  "Addison's",
  "Adela",
  "Adela's",
  "Adelaide",
  "Adelaide's",
  "Adele",
  "Adele's",
  "Adeline",
  "Adeline's",
  "Aden",
  "Aden's",
  "Adenauer",
  "Adenauer's",
  "Adhara",
  "Adhara's",
  "Adidas",
  "Adidas's",
  "Adirondack",
  "Adirondack's",
  "Adirondacks",
  "Adirondacks's",
  "Adkins",
  "Adkins's",
  "Adler",
  "Adler's",
  "Adm",
  "Admiralty",
  "Adolf",
  "Adolf's",
  "Adolfo",
  "Adolfo's",
  "Adolph",
  "Adolph's",
  "Adonis",
  "Adonis's",
  "Adonises",
  "Adrenalin",
  "Adrenalin's",
  "Adrenalins",
  "Adrian",
  "Adrian's",
  "Adriana",
  "Adriana's",
  "Adriatic",
  "Adriatic's",
  "Adrienne",
  "Adrienne's",
  "Advent",
  "Advent's",
  "Advents",
  "Adventist",
  "Adventist's",
  "Adventists",
  "Advil",
  "Advil's",
  "Aegean",
  "Aegean's",
  "Aelfric",
  "Aelfric's",
  "Aeneas",
  "Aeneas's",
  "Aeneid",
  "Aeneid's",
  "Aeolus",
  "Aeolus's",
  "Aeroflot",
  "Aeroflot's",
  "Aeschylus",
  "Aeschylus's",
  "Aesculapius",
  "Aesculapius's",
  "Aesop",
  "Aesop's",
  "Afghan",
  "Afghan's",
  "Afghans",
  "Afghani",
  "Afghani's",
  "Afghanistan",
  "Afghanistan's",
  "Afr",
  "Africa",
  "Africa's",
  "African",
  "African's",
  "Africans",
  "Afrikaans",
  "Afrikaans's",
  "Afrikaner",
  "Afrikaner's",
  "Afrikaners",
  "Afro",
  "Afro's",
  "Afros",
  "Afrocentric",
  "Afrocentrism",
  "Afrocentrism's",
  "Ag",
  "Ag's",
  "Agamemnon",
  "Agamemnon's",
  "Agana",
  "Agassi",
  "Agassi's",
  "Agassiz",
  "Agassiz's",
  "Agatha",
  "Agatha's",
  "Aggie",
  "Aggie's",
  "Aglaia",
  "Aglaia's",
  "Agnes",
  "Agnes's",
  "Agnew",
  "Agnew's",
  "Agni",
  "Agni's",
  "Agra",
  "Agra's",
  "Agricola",
  "Agricola's",
  "Agrippa",
  "Agrippa's",
  "Agrippina",
  "Agrippina's",
  "Aguascalientes",
  "Aguilar",
  "Aguilar's",
  "Aguinaldo",
  "Aguinaldo's",
  "Aguirre",
  "Aguirre's",
  "Agustin",
  "Agustin's",
  "Ahab",
  "Ahab's",
  "Ahmad",
  "Ahmad's",
  "Ahmadabad",
  "Ahmadabad's",
  "Ahmadinejad",
  "Ahmadinejad's",
  "Ahmed",
  "Ahmed's",
  "Ahriman",
  "Ahriman's",
  "Aida",
  "Aida's",
  "Aiken",
  "Aiken's",
  "Aileen",
  "Aileen's",
  "Aimee",
  "Aimee's",
  "Ainu",
  "Ainu's",
  "Airedale",
  "Airedale's",
  "Airedales",
  "Aires",
  "Aires's",
  "Aisha",
  "Aisha's",
  "Ajax",
  "Ajax's",
  "Akbar",
  "Akbar's",
  "Akhmatova",
  "Akhmatova's",
  "Akihito",
  "Akihito's",
  "Akita",
  "Akita's",
  "Akiva",
  "Akiva's",
  "Akkad",
  "Akkad's",
  "Akron",
  "Akron's",
  "Al",
]
`;

exports[`HunspellReader read dictionaries reads words with info from "nl" 1`] = `
Array [
  "aambeeld",
  "aambeeld-",
  "aambeelden",
  "aambeelden-",
  "aambei",
  "aambei-",
  "aambeien",
  "aambeien-",
  "aamborstig",
  "aamborstige",
  "aamborstigheid",
  "aamborstigheids",
  "aamborstigheids-",
  "aanaard",
  "aanaarding",
  "aanaardingen",
  "aanaardingen-",
  "aanaardings",
  "aanaardings-",
  "aanaardploeg",
  "aanaardploegen",
  "aanaardploegen-",
  "aanbaksel",
  "aanbaksels",
  "aanbeeldsblok",
  "aanbeeldsblokken",
  "aanbeeldsblokken-",
  "aanbeeld",
  "aanbeelden",
  "aanbeelden-",
  "aanbelang",
  "aanbelangen",
  "aanbelangen-",
  "aanberming",
  "aanbermingen",
  "aanbermingen-",
  "aanbermings",
  "aanbermings-",
  "aanbesteder",
  "aanbesteders",
  "aanbesteding",
  "aanbestedingen",
  "aanbestedingen-",
  "aanbestedings",
  "aanbestedings-",
  "aanbetaling",
  "aanbetalingen",
  "aanbetalingen-",
  "aanbetalings",
  "aanbetalings-",
  "aanbevelenswaardig",
  "aanbevelenswaardige",
  "aanbevelenswaardiger",
  "aanbeveling",
  "aanbevelingen",
  "aanbevelingen-",
  "aanbevelings",
  "aanbevelings-",
  "aanbevolenen",
  "aanbiddelijk",
  "aanbiddelijke",
  "aanbiddelijker",
  "aanbiddelijkst",
  "aanbiddelijkste",
  "alleraanbiddelijkst",
  "alleraanbiddelijkste",
  "aanbidder",
  "aanbidders",
  "aanbidding",
  "aanbiddingen",
  "aanbiddingen-",
  "aanbiddings",
  "aanbiddings-",
  "aanbidster",
  "aanbidsters",
  "aanbied",
  "aanbied-",
  "aanbieder",
  "aanbieders",
  "aanbieders-",
  "aanbieding",
  "aanbiedingen",
  "aanbiedingen-",
  "aanbiedings",
  "aanbiedings-",
  "aanblazing",
  "aanblazingen",
  "aanblazingen-",
  "aanbod",
  "aanbod-",
  "aanbodgericht",
  "aanbodgerichte",
  "aanbouwing",
  "aanbouwingen",
  "aanbouwingen-",
  "aanbouwsel",
  "aanbouwsels",
  "aanbreng",
  "aanbreng-",
  "aanbrenger",
  "aanbrengers",
  "aanbrengst",
  "aanbrengsten",
  "aanbrengsten-",
  "aanbrug",
  "aanbruggen",
  "aanbruggen-",
  "aandacht",
  "aandacht-",
  "aandachts",
  "aandachts-",
  "aandachtig",
  "aandachtige",
  "aandachtiger",
  "aandachtigere",
  "aandachtigst",
  "aandachtigste",
  "alleraandachtigst",
  "alleraandachtigste",
  "aandachtigheid",
  "aandachtstreep",
  "aandachtstreepje",
  "aandachtstreepjes",
  "aandachtstreepjes-",
  "aandachtstrepen",
  "aandachtstrepen-",
  "aandachttrekker",
  "aandachttrekkers",
  "aandaken",
  "aandeel",
  "aandeel-",
  "aandeeltje",
  "aandeeltjes",
  "aandeeltjes-",
  "aandelen",
  "aandelen-",
  "aandeelhoudend",
  "aandeelhoudende",
  "aandenken",
  "aandenkens",
  "Aandewiel",
  "Aandewiels",
  "aandien",
  "aandijking",
  "aandijkingen",
  "aandijkingen-",
  "aandijkings",
  "aandijkings-",
  "aandoening",
  "aandoeningen",
  "aandoeningen-",
  "aandoenings",
  "aandoenings-",
  "aandoenlijk",
  "aandoenlijke",
  "aandoenlijker",
  "aandoenlijks",
  "aandoenlijkst",
  "aandoenlijkste",
  "alleraandoenlijkst",
  "alleraandoenlijkste",
  "aandoenlijkheid",
  "aandraag",
  "aandraag-",
  "aandrager",
  "aandragers",
  "aandrang",
  "aandrang-",
  "aandrift",
  "aandrift-",
  "aandriften",
  "aandriften-",
  "aandrijfas",
  "aandrijfas-",
  "aandrijfassen",
  "aandrijfassen-",
  "aandrijf",
  "aandrijf-",
  "aandrijver",
  "aandrijvers",
  "aandrijving",
  "aandrijvingen",
  "aandrijvingen-",
  "aandrijvings",
  "aandrijvings-",
  "aandringer",
  "aandringers",
  "aandring",
  "aandringen",
  "aandringen-",
  "aandruk",
  "aandruk-",
  "aandrukrol",
  "aandrukrollen",
  "aandrukrollen-",
  "aandrukrolletje",
  "aandrukrolletjes",
  "aandrukrolletjes-",
  "aanduiding",
  "aanduidingen",
]
`;
